File: algorithms/114. Flatten Binary Tree to Linked List/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        while ( root ) {
            if ( root->left ) {
                auto tmp = root->left;
                while ( tmp && tmp->right ) {
                    tmp = tmp->right;
                }
                tmp->right = root->right;
                root->right = root->left;
                root->left = NULL;
            }
            root = root->right;
        }
    }
};

File: algorithms/114. Flatten Binary Tree to Linked List/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        def dfs(root):
            if root:
                dfs(root.right)
                dfs(root.left)
                root.left = None
                root.right = pre[0]
                pre[0] = root
        pre = [None]
        dfs(root)
